internal/controller/apiextensions/claim/reconciler.go

```diff
@@ -309,6 +309,9 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 				record.Event(cm, event.Warning(reasonBind, err))
 				return reconcile.Result{RequeueAfter: aShortWait}, nil
 			}
+			// If the resource is not found, cp is empty. But we need its name
+			// so that deletion logic can process the deletion.
+			cp.SetName(ref.Name)
```

Comment:
Could we instead gate the `r.client.Delete(ctx, cp)` so that it only runs `if meta.WasCreated(cp)` (i.e. if `cp` has a creation timestamp, which implies it has a name). That might be a little more obvious to readers.

Reply:
Hmm it reads like `delete only if it was created` but it might have been created and deleted earlier, so for the case where composite is already gone, `meta.WasCreated` sounds like an indirect way of checking whether `Get` succeeded. On the other hand, the place we call `cp.SetName` is very specific to this case as explained in the comment of the `if !kerrors.IsNotFound(err) || !meta.WasDeleted(cm)` check, so it felt more obvious to me as continuation of that logic. Also it feels kind of odd to both check `meta.WasCreated` and also ignore `NotFound`. I thought of removing `resource.IgnoreNotFound` but it's kind of a safety net there.

But both would cover all the cases we want so if you feel strongly, I can change it.

Reply:
> Also it feels kind of odd to both check meta.WasCreated and also ignore NotFound. I thought of removing resource.IgnoreNotFound but it's kind of a safety net there.

I kind of see where you're coming from here, but I don't think this is too weird as long as folks understand the two data sources. I feel that the "meta" in `meta.WasCreated` communicates that our belief that this resource exists is derived from our current client-side view of its metadata, whereas `resource.IgnoreNotFound` is a directive with regard to what to do about fresher data returned from the API server. e.g. "I want to delete this resource because I _think_ it exists, but it's okay if the delete fails because it doesn't."

Reply:
Agreed a `meta.Exists(o)` or something that talks about the current time would be handy. Switched the implementation to use `meta.WasCreated(o)`.